src/fsharp/Optimizer.fs

```diff
@@ -2595,21 +2598,23 @@ and OptimizeApplication cenv env (f0,f0ty,tyargs,args,m) =
 
     let shapes = 
         match f0' with 
-        | Expr.Val(vref,_,_) when Option.isSome vref.ValReprInfo -> 
-            let (ValReprInfo(_kinds,detupArgsL,_)) = Option.get vref.ValReprInfo 
-            let nargs = (args.Length) 
-            let nDetupArgsL = detupArgsL.Length
-            let nShapes = min nargs nDetupArgsL 
-            let detupArgsShapesL = 
-                List.take  nShapes detupArgsL |> List.map (fun detupArgs -> 
-                    match detupArgs with 
-                    | [] | [_] -> UnknownValue
-                    | _ -> TupleValue(Array.ofList (List.map (fun _ -> UnknownValue) detupArgs))) 
-            detupArgsShapesL @ List.replicate (nargs - nShapes) UnknownValue
-            
-        | _ -> args |> List.map (fun _ -> UnknownValue) 
-
-    let args',arginfos = OptimizeExprsThenReshapeAndConsiderSplits cenv env (List.zip shapes args) 
+        | Expr.Val(vref,_,_) ->
```

Comment:
@forki, I actually prefer the when clause on this pattern match it eliminates the duplication.

Does that make sense?

Kevin